README.md

```diff
@@ -15,10 +15,10 @@ the following cases:
 In all cases, a Pulse installation (or the Pulse repository clone) is
 laid out as follows:
 
-* in `lib/pulse`:
-  * the theoretical foundations of Pulse, in `core` (namespace `PulseCore`)
-  * the interface powering the Pulse DSL, `Pulse.Main.fsti`
-  * the Pulse standard library, in `lib` (namespace `Pulse.Lib`), with
+* in `lib/`:
+  * the theoretical foundations of Pulse, in `core/` (namespace `PulseCore`)
+  * the interface powering the Pulse DSL, `common/Pulse.Main.fsti`
+  * the Pulse standard library, in `pulse/` (namespace `Pulse.Lib`), with
     some subdirectories for typeclasses (in `class`, namespace
```

Comment:
Thanks Guido for your proposal!

With this PR, do lines 7-13 still hold?
> This repository has been designed to closely follow the Filesystem
Hierarchy Standard (FHS), so that it can be used the same way in all
the following cases:
> * direct use of the Pulse repository clone
> * installation into a FHS-like hierarchy (e.g. `/usr/local`)
> * installation as an _opam_ package, using the _opam_ package manager for
  OCaml

From what I understand, for this to hold, only the Pulse standard library would be installed via `make install`, and elaborated code produced by the Pulse elaborator should not depend on `lib/common`. Is my understanding correct? If so, this should be documented somewhere.

If that does not hold, how should a library developer depending on Pulse configure the location of the Pulse library in their Makefiles and how should they detect whether their users installed Pulse with opam or `make install` or clone it from the GitHub repository?

Reply:
`lib/common` is also installed (it must be since Pulse.Lib.Core.fsti is there). What we skip is installing PulseCore implementation files (though the checked files are installed... this is the case now too, and wasteful, so I'll remove it in a later PR).

All a user has to do is `--include /xyz/lib/pulse`, where `/xyz` is the root (PREFIX) of the installation. If they just build the repo without installing it, this is `out/lib/pulse``. If they make install, it is `/usr/local/lib/pulse`. And for OPAM it's `/home/user/.opam/sw/lib/pulse`.

So they do not really have to distinguish these cases at all, just point to the right `lib/pulse` directory, like until now? With the change that now there is an extra `out/` component for the local case.

I changed the README slightly. One thing I removed is `share/pulse/Makefile.include`, do we still want this? Given that the only flag users need is a single `--include`, I don't see why we want Pulse-specific Makefiles.

Reply:
> `lib/common` is also installed [...]
> All a user has to do is `--include /xyz/lib/pulse`, where `/xyz` is the root (PREFIX) of the installation. If they just build the repo without installing it, this is `` out/lib/pulse``. If they make install, it is ``/usr/local/lib/pulse`. And for OPAM it's `/home/user/.opam/sw/lib/pulse`.

1. does this mean that this single `--include` is enough for users, and `fstar.include` files from there take care of `include`-ing `lib/common`?
2. where is `lib/common` installed? I believe it cannot be installed as `/home/user/.opam/sw/lib/common` (and similarly for `make install`) and it has to go under `lib/pulse` instead

> With the change that now there is an extra out/ component for the local case.

Thanks Guido for clarifying this case. Then, I think it would be simpler to remove that now confusing "FHS" introductory paragraph from README.md: the structure of the repository now no longer matters since people using the repository should use the `out` subdirectory instead of the root directory of the repository.

>  One thing I removed is `share/pulse/Makefile.include`, do we still want this? Given that the only flag users need is a single `--include`, I don't see why we want Pulse-specific Makefiles.

Don't we also have `--load_cmxs pulse` ?

Reply:
> 1. does this mean that this single `--include` is enough for users, and `fstar.include` files from there take care of `include`-ing `lib/common`?

Sort of, but indeed the single include is enough. The file `lib/pulse/fstar.include` contains only `lib`, but everything is there in `lib/pulse/lib`. The directories `common` and `pulse` are flattened in the package, users don't ever see them, it's a convention of the source repo, not the package.

> 2. where is `lib/common` installed? I believe it cannot be installed as `/home/user/.opam/sw/lib/common` (and similarly for `make install`) and it has to go under `lib/pulse` instead

It all goes into `lib/pulse/lib`. I suppose a better name could be `lib/pulse/fstar`, since this directory has all F*-specific files (fst/fsti/checked). The parent `lib/pulse` has the built OCaml files, including `pulse.cmxs`.
 
> > With the change that now there is an extra out/ component for the local case.
> 
> Thanks Guido for clarifying this case. Then, I think it would be simpler to remove that now confusing "FHS" introductory paragraph from README.md: the structure of the repository now no longer matters since people using the repository should use the `out` subdirectory instead of the root directory of the repository.

Sounds good!

> 
> > One thing I removed is `share/pulse/Makefile.include`, do we still want this? Given that the only flag users need is a single `--include`, I don't see why we want Pulse-specific Makefiles.
> 
> Don't we also have `--load_cmxs pulse` ?

This is not needed since `#lang-pulse` will trigger F* to find the `pulse.cmxs` plugin in the include path and automatically load it.